### PR TITLE
httproute: pluralize hostname and allow for wildcards

### LINKED ISSUE #80: Use case: succinctly expose the same service on multiple names
**What would you like to be added**:

Allow multiple HTTP hostnames to be attached to the same set of rules.

**Why is this needed**:

See [this comment](https://github.com/kubernetes-sigs/service-apis/pull/43#discussion_r367145600).

> There’s a use case for having multiple hostnames in one endpoint, for example multi language site through example.TLD.
> A client can select to one or the other hostname and the developer wants not to duplicate her Rules. If we singularize we need to duplicate the Rules.

### PR DESCRIPTION
**Pull request by hbagdi**
This commit bundles two changes:
- pluralizes hostnames to define the same route
- allows for wildcards in the left-most label (same as Ingress V1)

Fix #80

### CODE REVIEW

**Comment by hbagdi** [comment]
While this PR has two different changes, I opened this up as one. Let me know if I should separate this out.

This PR also introduces some inconsistencies in how "wildcard" or "domain" (@jpeach  :D) matching is defined at the listener level vs the HTTPRoute level.

File: api/v1alpha1/httproute_types.go [hunk 1]
```diff
@@ -31,26 +31,38 @@ type HTTPRouteSpec struct {
 	Default *HTTPRouteHost `json:"default" protobuf:"bytes,2,opt,name=default"`
 }
 
-// HTTPRouteHost is the configuration for a given host.
+// HTTPRouteHost is the configuration for a given host(s).
```
Comment (robscott) [inline api/v1alpha1/httproute_types.go:34]:
```suggestion
// HTTPRouteHost is the configuration for a given set of hosts.
```

File: api/v1alpha1/httproute_types.go [hunk 2]
```diff
@@ -31,26 +31,38 @@ type HTTPRouteSpec struct {
...
-	// rules. For example, if the request header contains host: foo.example.com,
-	// an HTTPRoute with hostname foo.example.com will match. However, an
-	// HTTPRoute with hostname example.com or bar.example.com will not match.
-	// If Hostname is unspecified, the Gateway routes all traffic based on
-	// the specified rules.
+	// Incoming requests are matched against the host before the
+	// HTTPRoute rules. If the host is unspecified, traffic is routed
+	// based on the HTTPRouteRules.
+	//
+	// Host can be "precise" which is a domain name without the terminating
```
Comment (robscott) [inline api/v1alpha1/httproute_types.go:49]:
```suggestion
	// Hostnames can be "precise" which is a domain name without the terminating
```

File: api/v1alpha1/httproute_types.go [hunk 3]
```diff
@@ -31,26 +31,38 @@ type HTTPRouteSpec struct {
...
 	// 1. IPs are not allowed.
 	// 2. The `:` delimiter is not respected because ports are not allowed.
 	//
-	// Incoming requests are matched against Hostname before processing HTTPRoute
-	// rules. For example, if the request header contains host: foo.example.com,
-	// an HTTPRoute with hostname foo.example.com will match. However, an
-	// HTTPRoute with hostname example.com or bar.example.com will not match.
-	// If Hostname is unspecified, the Gateway routes all traffic based on
-	// the specified rules.
+	// Incoming requests are matched against the host before the
```
Comment (robscott) [inline api/v1alpha1/httproute_types.go:45]:
```suggestion
	// Incoming requests are matched against the hosts before the
```

**Comment by bowei** [comment]
looks like this is easy to fix and get merged


/approve

**Comment by hbagdi** [comment]
@robscott I've incorporated your feedback. PTAL.

**Pull request merged**